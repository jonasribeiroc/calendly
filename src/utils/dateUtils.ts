import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
    getDay,
    lastDayOfMonth,
    isSameMonth as isSameMonthFns,
    startOfMonth,
    endOfMonth,
    getDate,
} from 'date-fns';

function getDateUTC(date?: Date): Date {
    const baseDate = date instanceof Date ? date : new Date();

    return new Date(Date.UTC(
        baseDate.getUTCFullYear(),
        baseDate.getUTCMonth(),
        baseDate.getUTCDate(),
        baseDate.getUTCHours(),
        baseDate.getUTCMinutes(),
        baseDate.getUTCSeconds(),
        baseDate.getUTCMilliseconds()
    ));
}

function isSameMonth(date1: Date, date2: Date): boolean {
    return isSameMonthFns(date1, date2);
}

function isValidDateString(str: string, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str, formatStr, getDateUTC());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDateFromString(str: string | null, dateFormat: string): Date | undefined {
    if (str && isValidDateString(str, [dateFormat])) {
        return getDateUTC(new Date(str.replace('-', ',')));
    }
}

function getDateString(date: Date, formatStr?: string): string {
    return format(date, formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    return getDate(lastDayOfMonth(date));
}

function getWeekDay(date: Date): number {
    return getDay(date);
}

function getWeekDayName(date: Date): string {
    return format(date, 'EEEE');
}

function getMonthDay(date: Date): number {
    return Number(format(date, 'd'));
}

function getTime(date: Date): string {
    return format(date, 'HH:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    const firstDay = startOfMonth(date);

    return getDateUTC(new Date(getYear(firstDay), getMonth(firstDay), getMonthDay(firstDay), 0, 0, 0, 0));
}

function getLastDayOfMonth(date: Date): Date {
    const lastDay = endOfMonth(date);

    return getDateUTC(new Date(getYear(lastDay), getMonth(lastDay), getMonthDay(lastDay), 23, 59, 59, 999));
}

function getAvailableDays(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameMonth(d, date))
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameMonth,
    isValidDateString,
    getDateUTC,
    getDateFromString,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getMonthDay,
    getTime,
    getFirstDayOfMonth,
    getLastDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}