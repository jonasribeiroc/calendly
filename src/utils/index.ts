import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
} from 'date-fns';

function isSameYearMonth(str1?: string | null, str2?: string | null): boolean {
    if (!str1 || !str2) return false;

    const [year1, month1] = str1.split('-');
    const [year2, month2] = str2.split('-');

    return year1 === year2 && month1 === month2;
}

function isValidDateStr(str: string | null, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str || '', formatStr, new Date());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDate(str?: string | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd'): Date {
    return !!str && isValidDateStr(str, ['yyyy-MM-dd', 'yyyy-MM']) ? parse(str, formatStr || 'yyyy-MM', new Date()) : new Date();
}

function getDateString(date?: Date | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd' | 'yyyy-MM-dd\'T\'HH:mm:ss' | "yyyy-MM-dd'T'HH:mm:ssxxx"): string {
    return format(date || new Date(), formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    const year = getYear(date);
    const month = getMonth(date);

    return new Date(year, month, 0).getDate();
}

function getWeekDay(date: Date): number {
    return date.getUTCDay();
}

function getWeekDayName(date: Date): string {
    return date.toLocaleString('en-US', { weekday: 'long' });
}

function getDay(date: Date): number  {
    return Number(date.toLocaleString('en-US', { day: 'numeric' }));
}

function getTime(date: Date): string {
    return format(date, 'hh:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    return new Date(date.getFullYear(), date.getMonth(), 1);
  }

function getAvailableDays(availableDateTimes: string[], date: Date): number[] {
    const daysSet = new Set<number>();
    const year = getYear(date);
    const month = getMonth(date);

    availableDateTimes.forEach((dateTime) => {
        const date = new Date(dateTime);
        const dateYear = getYear(date);
        const dateMonth = getMonth(date);

        if (dateYear === year && dateMonth === month) {
            daysSet.add(getDay(date));
        }
    });

    return Array.from(daysSet).sort((a, b) => a - b);
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameYearMonth,
    isValidDateStr,
    getDate,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getDay,
    getTime,
    getFirstDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}