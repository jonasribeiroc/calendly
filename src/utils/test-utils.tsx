import React, { ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

interface AllTheProvidersProps {
  children: ReactNode;
}

const AllTheProviders = ({ children }: AllTheProvidersProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
