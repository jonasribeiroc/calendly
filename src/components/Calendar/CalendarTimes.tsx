import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { getMonthDay, getMonthName, getTime, getWeekDayName } from '../../utils/dateUtils';
import { Column } from '../Column';
import { Text } from '../Text';
import { Row } from '../Row';

const CalendarTimesContainer = styled.div`
    justify-items: center;
    width: 100%;
`;

const CalendarTimesTitle = styled(Text)`
    font-size: 15px;
    margin-bottom: 20px;
`;

const CalendarTimesColumn = styled(Column)`
    display: flex;
    gap: 10px;
    width: 90%;
`;

const CalendarTimesRow = styled(Row)`
    display: flex;
    width: 100%;
    gap: 6px;
`;

interface CalendarTimesButtonProps {
    $isSelected?: boolean 
}

const CalendarTimesButton = styled.button<CalendarTimesButtonProps>`
    height: 50px;
    border: 1px solid ${({ theme }) => theme.colors.primary.mendium};
    background: ${({ $isSelected, theme }) => ($isSelected ? theme.colors.primary.medium : theme.colors.white)};
    color: ${({ $isSelected, theme }) => ($isSelected ? theme.colors.white : theme.colors.primary.medium)};
    border-radius: 4px;
    cursor: pointer;
    text-align: center;
    font-weight: bold;
    width: ${({ $isSelected }) => ($isSelected ? '50%' : '100%')};
    font-size: 15px;
    transition: all 0.3s ease;

    &:hover {
        border: 2px solid ${({ theme }) => theme.colors.primary.medium};
    }

    &:disabled {
        background-color: #717171;
        border: none;
        cursor: default;
    }
`;

const SubmitButton = styled.button`
    height: 50px;
    background-color: ${({ theme }) => theme.colors.primary.medium};
    color: #fff;
    font-weight: bold;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    text-align: center;
    width: 50%;
    font-size: 15px;

    &:hover {
        background-color: ${({ theme }) => theme.colors.primary.mediumDark};
    }
`;

interface CalendarTimesProps {
    selectedDate: Date;
    availableTimes: Date[];
    onClick: (date: Date) => void;
}

export const CalendarTimes: React.FC<CalendarTimesProps> = ({
    selectedDate,
    availableTimes,
    onClick,
}) => {
    const [currentDate, setCurrentDate] = useState<Date | null>(null);

    const handleTimeClick = (date: Date) => {
        setCurrentDate(date);
    };

    const handleSubmit = () => {
        if (currentDate) {
            onClick(currentDate);
        }
    };

    useEffect(() => {
        setCurrentDate(null);
    }, [selectedDate]);

    return (
        <CalendarTimesContainer>
            <CalendarTimesTitle>
                {getWeekDayName(selectedDate)}, {getMonthName(selectedDate)} {getMonthDay(selectedDate)}
            </CalendarTimesTitle>

            <CalendarTimesColumn>
                {availableTimes.map((date: Date, index: number) => (
                    <CalendarTimesRow key={index}>
                        <CalendarTimesButton
                            $isSelected={currentDate === date}
                            onClick={() => handleTimeClick(date)}
                            disabled={currentDate === date}
                        >
                            {getTime(date)}
                        </CalendarTimesButton>
                        {currentDate && currentDate === date &&(
                            <SubmitButton onClick={handleSubmit}>
                                Next
                            </SubmitButton>
                        )}
                    </CalendarTimesRow>
                ))}
            </CalendarTimesColumn>


        </CalendarTimesContainer>
    );
};
