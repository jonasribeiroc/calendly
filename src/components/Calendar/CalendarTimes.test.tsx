import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import { CalendarTimes } from './CalendarTimes';
import * as dateUtils from '../../utils/dateUtils';
import { render } from '../../utils/test-utils';

jest.mock('../../utils/dateUtils');

describe('CalendarTimes Component', () => {
  const mockedDateUtils = dateUtils as jest.Mocked<typeof dateUtils>;

  const mockOnClick = jest.fn();

  const mockSelectedDate = new Date('2025-04-15T00:00:00Z');
  const mockAvailableTimes = [
    new Date('2025-04-15T10:00:00Z'),
    new Date('2025-04-15T14:00:00Z'),
    new Date('2025-04-15T16:30:00Z'),
  ];

  beforeEach(() => {
    jest.clearAllMocks();

    mockedDateUtils.getMonthDay.mockImplementation((date: Date) => date.getUTCDate());

    mockedDateUtils.getMonthName.mockImplementation((date: Date) => {
      return date.toLocaleString('default', { month: 'long' });
    });

    mockedDateUtils.getWeekDayName.mockImplementation((date: Date) => {
      return date.toLocaleString('default', { weekday: 'long' });
    });

    mockedDateUtils.getTime.mockImplementation((date: Date) => {
      return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    });
  });

  it('renders the correct title based on selectedDate', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const title = screen.getByText(
      `${mockedDateUtils.getWeekDayName(mockSelectedDate)}, ${mockedDateUtils.getMonthName(
        mockSelectedDate
      )} ${mockedDateUtils.getMonthDay(mockSelectedDate)}`
    );

    expect(title).toBeInTheDocument();
  });

  it('renders all available times as buttons', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    mockAvailableTimes.forEach((time) => {
      const timeButton = screen.getByText(mockedDateUtils.getTime(time));
      expect(timeButton).toBeInTheDocument();
      expect(timeButton).toBeEnabled();
    });
  });

  it('does not render SubmitButton when no time is selected', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const submitButtons = screen.queryAllByText('Next');
    expect(submitButtons).toHaveLength(0);
  });

  it('selects a time when a time button is clicked and shows SubmitButton', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const timeToSelect = mockAvailableTimes[0];
    const timeButton = screen.getByText(mockedDateUtils.getTime(timeToSelect));
    
    fireEvent.click(timeButton);

    expect(timeButton).toBeDisabled();

    const submitButton = screen.getByText('Next');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toBeEnabled();
  });

  it('calls onClick with the selected date when SubmitButton is clicked', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const timeToSelect = mockAvailableTimes[1];
    const timeButton = screen.getByText(mockedDateUtils.getTime(timeToSelect));
    
    fireEvent.click(timeButton);

    const submitButton = screen.getByText('Next');
    fireEvent.click(submitButton);

    expect(mockOnClick).toHaveBeenCalledWith(timeToSelect);
  });

  it('only one SubmitButton is visible at a time', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const firstTimeButton = screen.getByText(mockedDateUtils.getTime(mockAvailableTimes[0]));
    fireEvent.click(firstTimeButton);

    const secondTimeButton = screen.getByText(mockedDateUtils.getTime(mockAvailableTimes[1]));
    fireEvent.click(secondTimeButton);

    const submitButtons = screen.getAllByText('Next');
    expect(submitButtons).toHaveLength(1);
  });

  it('resets selected time when selectedDate changes', () => {
    const { rerender } = render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const timeButton = screen.getByText(mockedDateUtils.getTime(mockAvailableTimes[2]));
    fireEvent.click(timeButton);

    const submitButton = screen.getByText('Next');
    expect(submitButton).toBeInTheDocument();

    const newSelectedDate = new Date('2025-04-16T00:00:00Z');
    rerender(
      <CalendarTimes
        selectedDate={newSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const newSubmitButton = screen.queryByText('Next');
    expect(newSubmitButton).not.toBeInTheDocument();

    expect(timeButton).toBeEnabled();
  });

  it('disables SubmitButton when no time is selected', () => {
    render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );

    const submitButton = screen.queryByText('Next');
    expect(submitButton).not.toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <CalendarTimes
        selectedDate={mockSelectedDate}
        availableTimes={mockAvailableTimes}
        onClick={mockOnClick}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
