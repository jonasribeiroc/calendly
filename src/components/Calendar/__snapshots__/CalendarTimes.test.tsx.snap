// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarTimes Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-fAUdSK joDQZj"
  >
    <div
      class="sc-gtLWhw sc-dntaoT jnpasD clCmLo"
    >
      Monday, April 15
    </div>
    <div
      class="sc-blHHSb sc-ivxoEo edipXs izTEag"
    >
      <div
        class="sc-egkSDF sc-ghWlax hEyCam bfpDIU"
        data-testid="07:00 AM"
      >
        <button
          class="sc-kLhKbu fvfsF"
        >
          07:00 AM
        </button>
      </div>
      <div
        class="sc-egkSDF sc-ghWlax hEyCam bfpDIU"
        data-testid="11:00 AM"
      >
        <button
          class="sc-kLhKbu fvfsF"
        >
          11:00 AM
        </button>
      </div>
      <div
        class="sc-egkSDF sc-ghWlax hEyCam bfpDIU"
        data-testid="01:30 PM"
      >
        <button
          class="sc-kLhKbu fvfsF"
        >
          01:30 PM
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
